src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs

```diff
@@ -378,6 +378,58 @@ public static void Equal<T>(HashSet<T> expected, HashSet<T> actual)
             }
         }
 
+        /// <summary>
+        /// Validates that the actual collection contains same items as expected collection. If the test fails, this will display:
+        /// 1. Count if two collection count are different;
+        /// 2. Missed expected collection item when found
+        /// </summary>
+        /// <param name="expected">The collection that <paramref name="actual"/> should contain same items as</param>
+        /// <param name="actual"></param>
+        /// <param name="comparer">The comparer used to compare the items in two collections</param>
+        public static void CollectionEqual<T>(IEnumerable<T> expected, IEnumerable<T> actual, IEqualityComparer<T> comparer)
+        {
+            var actualItemCountMapping = new Dictionary<T, ItemCount>(comparer);
+            var actualCount = 0;
+            foreach (T actualItem in actual)
+            {
+                if (actualItemCountMapping.TryGetValue(actualItem, out ItemCount countInfo))
+                {
+                    countInfo.Original++;
+                    countInfo.Remain++;
+                }
+                else
+                {
+                    actualItemCountMapping[actualItem] = new ItemCount(1, 1);
+                }
+                
+                actualCount++;
+            }
+
+            var expectedArray = expected.ToArray();
+            var expectedCount = expectedArray.Length;
+
+            if (expectedCount != actualCount)
+            {
+                throw new XunitException($"Expected count: {expectedCount}{Environment.NewLine}Actual count: {actualCount}");
+            }
+
+            for (var i = 0; i < expectedCount; i++)
+            {
+                T currentExpectedItem = expectedArray[i];
+                if (!actualItemCountMapping.TryGetValue(currentExpectedItem, out ItemCount countInfo))
+                {
+                    throw new XunitException($"Expected: {currentExpectedItem} but Not found");
+                }
+
+                if (countInfo.Remain == 0)
+                {
+                    throw new XunitException($"Expected more: {currentExpectedItem}{Environment.NewLine}Actual: totally {countInfo.Original} {currentExpectedItem}");
```

Comment:
The message could be clearer. I think it's enough to mention that the collections are not equal.

Reply:
Fixed by refining exception message, but still keep some non-matched count info because I feel this issue is just to try resolve similar situation. Thanks @maryamariyan !

Excited to see you @maryamariyan have approved it :) But not sure who should be able to merge it as seems I have no write permission to do that.